pkg/binding/bindingpolicy-resolution-broker.go

```diff
@@ -0,0 +1,129 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package binding
+
+import (
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/abstract"
+	"github.com/kubestellar/kubestellar/pkg/util"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sync"
+)
+
+// ResolutionBroker allows for the registration of callback functions that
+// are called when a resolution is updated.
+// It also allows for the retrieval of resolutions for a given bindingPolicyKey.
+type ResolutionBroker interface {
+	// RegisterCallback adds a new callback function that will be called,
+	// on a separate goroutine, when a resolution is updated. The callback
+	// function should accept a bindingPolicyKey.
+	// There is no deduplication of callbacks.
+	RegisterCallback(func(bindingPolicyKey string))
+
+	// GetResolution retrieves the resolution for a given bindingPolicyKey.
+	// If no resolution is associated with the given key, nil is returned.
+	GetResolution(bindingPolicyKey string) *Resolution
+}
+
+// Resolution is a struct that represents the resolution of a binding policy.
+// It contains the destinations and object data for the resolution.
+type Resolution struct {
+	// Destinations is a list of destinations that are the "where" part of the
+	// resolution.
+	Destinations []v1alpha1.Destination
+	// ObjectIdentifierToData is a map of object identifiers to their
+	// corresponding ObjectData.
+	ObjectIdentifierToData map[util.ObjectIdentifier]struct {
+		ResourceVersion  string
+		StatusCollectors []string
+	}
+}
+
+// NewResolutionBroker creates a new ResolutionBroker with the given
+// bindingPolicyResolutionGetter function.
+// The latter function's returned `bindingPolicyResolution` is expected to be
+// thread-safe under the constraint that only its methods are used.
+func NewResolutionBroker(bindingPolicyResolutionGetter func(bindingPolicyKey string) *bindingPolicyResolution) ResolutionBroker {
+	return &resolutionBroker{
+		bindingPolicyResolutionGetter: bindingPolicyResolutionGetter,
+	}
+}
+
+// resolutionBroker implements the ResolutionBroker interface.
+// The broker requires a bindingPolicyResolutionGetter function that retrieves
+// resolutions from a BindingPolicyResolver.
+type resolutionBroker struct {
+	sync.Mutex
+	callbacks []func(bindingPolicyKey string)
+	// bindingPolicyResolutionGetter is a function that retrieves the
+	// resolution for a given bindingPolicyKey. If no resolution is associated
+	// with the given key, nil is returned.
+	// The returned `bindingPolicyResolution` is expected to be thread-safe as
+	// long as only its methods are used.
+	bindingPolicyResolutionGetter func(bindingPolicyKey string) *bindingPolicyResolution
+}
+
+// RegisterCallback adds a new callback function that will be called, on a
+// separate goroutine, when a resolution is updated. The callback function
```

Comment:
Would the caller be simpler if there were also an initial callback?

Reply:
This also makes sense. I thought about registering callbacks before "starting" the resolver - I will go with the simplest of the two when the broker is used.